runtime-light/component/component.cpp

```diff
@@ -6,18 +6,104 @@
+  component_ctx.php_script_mutable_globals_singleton.get_superglobals().v$_POST = f$json_decode(http_query, true);
+}
+
+task_t<uint64_t> init_kphp_cli_component() noexcept {
+  co_return co_await wait_for_incoming_stream_t{};
+}
+
+task_t<uint64_t> init_kphp_server_component() noexcept {
+  uint32_t magic{};
+  const auto stream_d{co_await wait_for_incoming_stream_t{}};
+  const auto read{co_await read_exact_from_stream(stream_d, reinterpret_cast<char *>(std::addressof(magic)), sizeof(uint32_t))};
+  php_assert(read == sizeof(uint32_t));
+  if (magic == K2_INVOKE_HTTP_MAGIC) {
+    const auto [buffer, size]{co_await read_all_from_stream(stream_d)};
+    init_http_superglobals(string{buffer, static_cast<string::size_type>(size)});
+    get_platform_context()->allocator.free(buffer);
+  } else if (magic == K2_INVOKE_JOB_WORKER_MAGIC) {
+    php_error("not implemented");
+  } else {
+    php_error("server got unexpected type of request: 0x%x", magic);
+  }
+
+  co_return stream_d;
+}
+
+int32_t merge_output_buffers(ComponentState &component_ctx) noexcept {
+  Response &response{component_ctx.response};
+  php_assert(response.current_buffer >= 0);
+
+  int32_t ob_first_not_empty{};
+  while (ob_first_not_empty < response.current_buffer && response.output_buffers[ob_first_not_empty].size() == 0) {
+    ++ob_first_not_empty; // TODO: optimize by precomputing final buffer's size to reserve enough space
+  }
+  for (auto i = ob_first_not_empty + 1; i <= response.current_buffer; i++) {
+    response.output_buffers[ob_first_not_empty].append(response.output_buffers[i].c_str(), response.output_buffers[i].size());
+  }
+  return ob_first_not_empty;
+}
+
+} // namespace
 
 void ComponentState::init_script_execution() noexcept {
   kphp_core_context.init();
-  init_php_scripts_in_each_worker(php_script_mutable_globals_singleton, main_task);
-  scheduler.suspend(std::make_pair(main_task.get_handle(), WaitEvent::Rechedule{}));
+  init_php_scripts_in_each_worker(php_script_mutable_globals_singleton, main_task_);
+  scheduler.suspend(std::make_pair(main_task_.get_handle(), WaitEvent::Rechedule{}));
+}
+
+template<ComponentKind kind>
+task_t<void> ComponentState::run_component_prologue() noexcept {
+  static_assert(kind != ComponentKind::Invalid);
+
+  component_kind_ = kind;
+  if constexpr (kind == ComponentKind::CLI) {
+    standard_stream_ = co_await init_kphp_cli_component();
+  } else if constexpr (kind == ComponentKind::Server) {
+    standard_stream_ = co_await init_kphp_server_component();
+  }
+}
+
+template task_t<void> ComponentState::run_component_prologue<ComponentKind::CLI>();
+template task_t<void> ComponentState::run_component_prologue<ComponentKind::Server>();
+template task_t<void> ComponentState::run_component_prologue<ComponentKind::Oneshot>();
+template task_t<void> ComponentState::run_component_prologue<ComponentKind::Multishot>();
+
+task_t<void> ComponentState::run_component_epilogue() noexcept {
```

Comment:
Why prologue is has compile time conditions. but epilogue has runtime?

Reply:
Prologue is only called by the compiler, that's why we can use compile-time expressions. At the same time, epilogue is called not only by the compiler, but, for example, by exit functions too.